### PR TITLE
Improve plugin API documentation 

### PR DESCRIPTION
**Pull request by fjetter**
I was made aware of the documentation of our plugins to be a bit misleading and outdated. This should reflect reality a bit better and add some links to relevant pages.

### CODE REVIEW

File: distributed/diagnostics/plugin.py [hunk 1]
```diff
@@ -96,6 +94,14 @@ def transition(
     ) -> None:
         """Run whenever a task changes state
 
+        For a description of the transition mechanism and the available states,
+        see :ref:`Scheduler task states <scheduler-task-state>`.
+
+        Note::
```
Comment (jacobtomlinson) [inline distributed/diagnostics/plugin.py:100]: Maybe this should be a warning?
```suggestion
        Warning::
```

**Review by jacobtomlinson — APPROVED** [review]
LGTM

File: distributed/diagnostics/plugin.py (hunk 1 above)
Comment (fjetter) [reply distributed/diagnostics/plugin.py:100]: yeah, I was not sure how much I wanted to scare people off... I personally would prefer for this API not even to exist so I guess a warning is indeed more appropriate :)

File: distributed/diagnostics/plugin.py [hunk 2]
```diff
@@ -96,6 +94,14 @@ def transition(
     ) -> None:
         """Run whenever a task changes state
 
+        For a description of the transition mechanism and the available states,
+        see :ref:`Scheduler task states <scheduler-task-state>`.
+
+        Warning::
```
Comment (jrbourbeau) [inline distributed/diagnostics/plugin.py:100]: This isn't rendering correctly in the built docs https://distributed--7653.org.readthedocs.build/en/7653/plugins.html

I think we want something like `.. warning::` instead to make sphinx happy (I've not built the docs locally with this suggestion though)

File: distributed/diagnostics/plugin.py [hunk 3]
```diff
@@ -177,13 +182,18 @@ def teardown(self, worker):
...
-        in the state of each task. The Worker owning the task is then notified of this
-        state transition.
+        Throughout the lifecycle of a task (see :doc:`Worker State
+        <worker-state>`), Workers are instructed by the scheduler to compute
+        certain tasks, resulting in transitions in the state of each task. The
+        Worker owning the task is then notified of this state transition.
 
         Whenever a task changes its state, this method will be called.
 
+        Warning::
```
Comment (jrbourbeau) [inline distributed/diagnostics/plugin.py:192]: Similar comment here

File: distributed/diagnostics/plugin.py [hunk 4]
```diff
@@ -24,13 +24,11 @@
...
     """Interface to extend the Scheduler
 
-    The scheduler operates by triggering and responding to events like
-    ``task_finished``, ``update_graph``, ``task_erred``, etc..
-
-    A plugin enables custom code to run at each of those same events.  The
-    scheduler will run the analogous methods on this class when each event is
-    triggered.  This runs user code within the scheduler thread that can
-    perform arbitrary operations in synchrony with the scheduler itself.
+    A plugin enables to run custom hooks when specific events occur.  The
```
Comment (jrbourbeau) [inline distributed/diagnostics/plugin.py:27]: Minor wording comment

```suggestion
    A plugin enables custom hooks to run when specific events occur.  The
```

**Review by jrbourbeau — COMMENTED** [review]
LGTM overall -- I left a few minor suggestions. @fjetter it's close to EOD for you, if you'd like me to push this over the finish line, just let me know

**Pull request merged**